support custom room counts from yaml

Rnd-Guy opened the pull request:
Updated 07/10/24
---
This top post will be kept up to date and should always provide a summary of the current state of the PR

### Adds custom room count as a yaml option, as well as agate knife percentage

In order to do this, the following were done:
* Connecting to the server is now forced before any gameplay begins, to get the room count before gen
* Added connection loading screens which also say when connection has failed
* Removed requirement for dungeon to have `max_dist >= 50`
* Room distance is multiplied by `50 / max_dist` for enemy scaling, audio, horde room treasure calculations. 
  * max_dist is reduced by ~3% to allow for a bit of overshooting, to mimic how vanilla sometimes goes above 50. This is to keep enemy proportions roughly similar
* Crystal scaling is increased by `10000 / total enemies` to balance the reduced gem count from having fewer enemies
  * If `gem_count * crystal_scaling < 250000` (which is roughly the minimum vanilla expects to get) then we redo generation, that way we don't get affected by the low lows.

### Testplays (of latest changes):
![373988851-a65e43a6-87bb-45c9-bf9a-b94e028418cd](https://github.com/user-attachments/assets/381bc778-736e-4938-b89a-eed39384c401)
Note that with these testplays, no crystal efficiency was picked up and no bosses were killed until the very end, so even vanilla cannot get 18 caches with this strategy hence it's not a consideration in this PR.

### Expected total gems from enemies after multiplier (calculated at gen time):
![373988367-cd1f9ad1-bd3d-48f8-8b7c-d6f71ac86a73](https://github.com/user-attachments/assets/52f46530-04e9-41c9-a289-b4baca59168a)

### Other notes:
* Testing has been limited to specifically 300 rooms, 750, and 1500 room counts (dubbed small, medium and large) to make it easier to balance correctly.
  * The min room count that does not crash is 15, but there's not enough horde rooms to ensure 18 chests. A safe min room count is 100.
* knife option is really OP. Should it even be an option? Should we hide it? Should it even exist on smaller dungeons?
* Should be backwards compatible
* Some stats were added to estimate how many gems can be picked up, which I used to estimate crystal calculations. These have been added to the stats screen but are not visible unless the flag is set to true in code.

Gif of finding agate knife at room 100, with small dungeon displayed in map:
(Note gif has a bug where if you pick up the knife then quit then picking the knife again does nothing but this has since been fixed)
![meritous-100-rooms](https://github.com/user-attachments/assets/7a256bfc-bb10-44ee-b5d3-4b23c66ef9a5)

### Changelog / Full test history
* A baseline vanilla test was done on 0.5.0, which is the main target to compare against
* First was a number of 100 room tests back when crystal scaling was done by room_count, which was giving too many crystals
* A full test (300, 750, 1500) was done after this was changed to enemy count, but showed great variance in crystal count. The biggest amplifier however was a bug where the multiplier affected crystals dropped from death, heavily inflating fewer riin count runs. Was subsequently fixed afterwards.
* Added average gem count by counting up the estimated gems dropped by each enemy spawned in the dungeon as actually playing through games takes a very long time
* Repeatedly genned dungeons and measured up the average gem count between the different sizes to try to match vanilla
  * 16 gens of small medium large and vanilla showed that currently medium and large were genning with too few crystals
  * Adjusted max_dist formula to max_dist - 2 to allow overshooting and tried again, now all sizes had too many crystals
  * Adjusted max_dist formula to max_dist * 0.95, and regenerated dungeon if calculated gems < 250000. Was a bit on the low side.
  * Did 50 more gens of vanilla to get an average overshoot distance, gem stats and enemy count
  * Adjusted max_dist formula to max_dist * 0.97, expected enemies to 10480. Was a bit too high
  * Adjusted expected enemies to 10300. Was a bit too high.
  * Adjusted expected enemies back to 10000. Seemed to be ok, however this was getting more crystals than the previous 95% test with 10000. Not sure why.
* Did another full test (300, 750, 1500) and looks much more ok

Rnd-Guy committed:
support custom room counts from yaml

Rnd-Guy commented:
In case it's helpful, I'll attach the apworld changes I've used to get this to work, feel free to change them as desired

`__init__.py`
```diff
@@ -57,6 +57,8 @@ class MeritousWorld(World):
        self.include_psi_keys = False
        self.item_cache_cost = 0
        self.death_link = False
+       self.rooms = 3000
+       self.agate_knife_percent = 100

    @staticmethod
    def _is_progression(name):


@@ -101,6 +103,8 @@ class MeritousWorld(World):
        self.include_psi_keys = self.options.include_psi_keys.value
        self.item_cache_cost = self.options.item_cache_cost.value
        self.death_link = self.options.death_link.value
+       self.rooms = self.options.rooms.value
+       self.agate_knife_percent = self.options.agate_knife_percent.value

    def create_regions(self):
        create_regions(self.multiworld, self.player)


@@ -177,5 +181,7 @@ class MeritousWorld(World):
        return {
            "goal": self.goal,
            "cost_scale": cost_scales[self.item_cache_cost],
-           "death_link": self.death_link
+           "death_link": self.death_link,
+           "rooms": self.rooms,
            "agate_knife_percent": self.agate_knife_percent,
        }
```

`Options.py`
```diff
from dataclasses import dataclass

import typing
-from Options import Option, DeathLink, Toggle, DefaultOnToggle, Choice, PerGameCommonOptions
+from Options import Option, DeathLink, Toggle, DefaultOnToggle, Choice, PerGameCommonOptions, Range, NamedRange


cost_scales = {


@@ -53,6 +53,37 @@ class ItemCacheCost(Choice):
    default = 0


+class Rooms(NamedRange):
+   """
+       Sets the total room count for the dungeon generated, which changes the length of the game.
+       Crystal gain will be multiplied to compensate for the reduced room count.
+
+       - **Original (default):** 3000 rooms, ie base game amount
+       - **Small:** 300 rooms
+       - **Medium:** 750 rooms
+       - **Large:** 1500 rooms
+   """
+   display_name = "Rooms"
+   range_start = 15
+   range_end = 3000
+   default = 3000
+   special_range_names = {
+       "original": 3000,
+       "small": 300,
+       "medium": 750,
+       "large": 1500,
+   }
+
+class AgateKnifePercent(Range):
+   """
+       How much % of rooms to discover to count as 100% room completion.
+       Once this % is reached, the agate knife will spawn
+   """
+   display_name = "Agate Knife %"
+   range_start = 0
+   range_end = 100
+   default = 100
+
@dataclass
class MeritousOptions(PerGameCommonOptions):
    goal: Goal

@@ -60,3 +91,5 @@ class MeritousOptions(PerGameCommonOptions):
    include_evolution_traps: IncludeEvolutionTraps
    item_cache_cost: ItemCacheCost
    death_link: DeathLink
+   rooms: Rooms
+   agate_knife_percent: AgateKnifePercent
```

Rnd-Guy commented:
Found a bug with the branch, same as the bug shown in the gif of this PR:
* If a special item (eg seal or agate knife) is picked up that was gotten already (eg getting item then quitting before saving) in the slot, then the special item may fail to be given locally. This can be quite serious if the cursed seal is unable to be picked up a second time. This is likely due to the reordering of the setup before gameplay begins but not sure what it is yet. Tested once in normal branch and it didn't happen

![meritous-no-cursed-seal](https://github.com/user-attachments/assets/143378ef-c5f1-4b1d-a7e1-e340fc3298e6)

Rnd-Guy commented:
Can indeed confirm this bug is indeed related to the order of operations for game start, possibly because by waiting for the connection we trigger the on_connect handlers too early.

The specific difference happens when we pick up the special item. We correctly identify that we need to pick it up, but we fail the `if` condition shown below whereas in normal branch we correctly go inside and receive our cursed seal.
```C
void InternalCollectAPItem(uint64_t location)
{
  if (apStores[location / 24]->HasItemStored(location % 24))
    ReceiveItem((t_itemTypes)(apStores[location / 24]->GetStoredItem(location % 24) % AP_OFFSET),
                storeNames[location / 24]);
  else {
    std::list<int64_t> check;
    check.push_back(location + AP_OFFSET);
    ap->LocationScouts(check);
    ap->LocationChecks(check);
  }
}
```

Through trial and error it looks like it's specifically the case for when the connection is successful before we call `ReadStoreData()`, which makes sense given it's store related.
However a lot of the load game functions cannot be reordered because map data needs to be read in a specific order. That is, `ReadStoreData()` must be called after `RandomGenerateMap()`, but connection must happen before `RandomGenerateMap()` and yet we need connection handlers to run after `ReadStoreData()`.

This bug is really punching above it's weight class here D:

Rnd-Guy commented:
My theory is that `ReadStoreData()` was resetting the store state, which is the second time it gets reset during initialisation (the first is during `InitStores()` which is usually after generate map but is now done at the start).

Normally this is harmless, as it gets reset twice before anything is added, but I think because we now finish connecting and handle connection handlers between the first and second store resets, funny business occurs. The bug no longer happens when I remove the second reset (ie remove `CreateAPStores()` below)

```
void ReadAPState()
{
  CreateAPStores();

  int progress = 0;
  nextCheckToGet = FRInt();
  for (auto store: apStores) {
    store->SetCostFactor(FRInt());
    for (size_t x = 0; x < store->GetLength(); x++) {
      auto collected = FRChar();
      if (collected) store->MarkCollected(x);
      auto hasItem = (bool)FRChar();
      if (hasItem) store->StoreItem(x, FRInt64());
    }
    UpdateLoadingScreen((float)(++progress + 1) / (IS_MAX + 1));
  }
```
It makes sense because
* The reason I had to move `InitStores` to the start is because the on_connect handlers need to connect with the stores, so it definitely does something in that area
* `ReadAPStores()` is only called on game load, and the bug doesn't happen on a new game.

However the theory is broken because that should mean I can replicate it by starting a new game on a slot that has picked up a a special item, save game, quit and load game (to trigger the bug) and then try picking up the item again but it's working fine. Thankfully this is intended behaviour so it's not too bad of a thing if I can no longer replicate it but leaves uncertainty.

For now I think I'll just remove the call to `CreateAPStores()` and just monitor if this is an issue during further testplay.

Rnd-Guy committed:
proposed fix for special item bug

Rnd-Guy commented:
Just realised crystal counts shouldn't be scaled by room count but instead by monster count. 
Potentially can scale it by 10000 / monster count, as base game usually creates just over 10000 monsters, though I've seen sub 10k before. 

For example, in the first gif above we have 100 rooms and 771 monsters. 30x less rooms gives me a crystal multiplier of 30 despite only having ~13x less monsters, off by more than a factor of two, which helps explain a bit of the surplus crystals that I'm noticing

Rnd-Guy committed:
change crystal gain to be based off enemy count

also adds fractional handling

Rnd-Guy commented:
Did some more testing but unfortunately looks like there's an issue with crystal counts 🙁

Here's a chart of my crystals on a few runs with this PR, compared to a 3000 room count run I did on the sep 2024 big async. 
To ensure my comparison runs are consistent:
* I plando'd crystal efficiency onto alpha chest 18 (new to plando so I couldn't figure out how to put it to 24), to match how I did not get this on my big async run.
* Only defeated bosses as the last few rooms, thus 0 evolutions.

![image](https://github.com/user-attachments/assets/d8a193fc-618d-4123-9260-f818f1564187)

* Naturally the large 1500 room count is a major issue and that's enough to block this PR from being mergeable. Maybe rounding issue causing values to be halved? Though I had added code to handle partial crystals...
* The medium 750 room may be similar to the 1500 room issue. 
* Small room count may be highly susceptible to variance (eg psi key room at edge of dungeon) but perhaps the horde rooms being a larger proportion of the overall dungeon is inflating their value to above normal? Probably needs more tests.

The three run sizes here (300, 750, 1500) were sorta arbitrary room counts that I chose to represent `small`, `medium` and `large` choices to be used as yaml options. Not much thought was particularly done to pick those numbers, but playtime wise these took about 40min, 1hr10 and 2 hours respectively to 100% which I think is really nice!
It was suggested that possibly as a first approach, we stick to a limited set of room sizes to make the first implementation easier.

As for other testing results, I've not encountered any other major issues. 
I've confirmed the "not picking up agate/cursed seal" issue doesn't happen anymore after removing the `CreateAPStores()` call for example.

Once I figure out the crystal count issues, I think I'll be happy with how it is, but for now I'll put this to draft to reflect how it's not ready to be merged in this current state.

Edit: Found a bug where the crystals we drop on death (we lose 1/3, then 95% of that is turned into crystals we can pick back up) also get multiplied so we gain back more than we lost. Whoops! That might explain why 300 rooms gained so many more crystals.

Rnd-Guy committed:
avoid multiplying gems dropped on player death

Rnd-Guy committed:
crystal balancing for lower room counts

Rnd-Guy commented:
Made some changes to the crystal scaling algorithm to try to match vanilla a bit more, as well as a couple failsafes.

**Summary**
The two main factors that currently govern how many crystals smaller dungeons get are:
* room proportions. Having more rooms considered "far" results in more harder enemy spawns which drop much more gems. 
* expected vanilla enemy count, used to calculate crystal scaling (ie `expected vanilla enemy count / total enemies`)

**Overshoot distance - aim to match room and thus enemy proportions**
Vanilla calculates stuff based on room distance from center, with a cap of 50. 
I've made the assumption that if a dungeon has a distance higher than 50, then it likely has more rooms considered "far away" enough to spawn harder enemies, which in turn affects gems dropped.
Therefore by letting smaller dungeons also overshoot a little, I can mimic this behaviour to keep enemy proportions similar to vanilla.

I genned 50 vanilla dungeons and got an average max distance of 51.4, so I've multiplied the max distance of smaller dungeons by 50/51.4 (~97.2%) to mimic this behaviour, to keep enemy proportions roughly the same as what vanilla might have.

**Min gem check - at least the minimum that vanilla can generate with**
I calculate the average number of gems dropped (for each monster add `(min+max)/2`) and then multiply that by the expected monster count. If this is below what vanilla can generate (~250000) then I redo the dungeon generation.
This then prevents the smaller dungeons from being too volatile and not genning with enough crystals.

Since this skews the average, I've kept the expected monster count to be 10000, when the average is more like 10480 for vanilla, to keep the smaller dungeons from being too volatile and genning with too many crystals
* Average of 98,99,100,101,102 and 80,90,100,110,120 is 100, but removing entries less than 98 skews the right average to now be 110 for example

Before:
![image](https://github.com/user-attachments/assets/4b2cda56-a1ba-435f-9dca-f6a331dd061c)

After:
![image](https://github.com/user-attachments/assets/cd1f9ad1-bd3d-48f8-8b7c-d6f71ac86a73)

After with testplays:
![image](https://github.com/user-attachments/assets/a65e43a6-87bb-45c9-bf9a-b94e028418cd)

This is significantly better than the first attempt!
The 750 and 1500 are still a tiny bit lower than I expected them to be but perhaps small sample size issue. I'm much more happier with this now.

Note these runs were done with the worst crystal strats (no efficiency, no bosses until the very end) and so even vanilla doesn't get 18 caches. As a result we only need to aim to be close enough to vanilla.

With this, I'm reopening this PR!

Rnd-Guy committed:
remove accidental nerf to crystal items

Rnd-Guy committed:
Revert "remove accidental nerf to crystal items"

This reverts commit eb37fa82fefade46e2b63e16dad68269d6bfe147.

Rnd-Guy committed:
Remove accidental nerf to crystal items

FelicitusNeko commented on src/mapgen.c:
```diff
@@ -41,6 +42,16 @@ int DoRepeat = 0;
+// vanilla does calculations based on a max dist of 50, but can gen dungeons with higher than 50.
+// we then mimic this overshooting for small dungeons too, to try to keep the proportion of difficult/far rooms the same.
+// generated 50 vanilla worlds and got their average max_dist (51.4) so for calculations we'll use 50/51.4 (ie 97.276%) of the actual max distance
+float max_dist_overshoot = 0.97276f; // 
```
This looks like it can be `const`

Rnd-Guy replied:
good idea!

FelicitusNeko commented on src/mapgen.h:
```diff
@@ -55,8 +55,10 @@ typedef struct {
+extern float dist_scaling;
 
 void RandomGenerateMap();
+void InitStores();
```
I'm not sure why we're declaring `InitStores()` here

Rnd-Guy replied:
Good spot, that must've been a leftover artifact from debugging something. Will remove

FelicitusNeko reviewed (commented):
Sorry for leaving this unreviewed for so long >__< Just a couple of things I wanted to point out.

Rnd-Guy committed:
PR comments - remove unused header fn + set const float

Rnd-Guy commented:
I've made the changes according to your review but right now I can't actually compile it to see if it works D:
I completely scuffed my c dev environment trying to figure out a good way to debug the game ^^;

I'll try to fix it this weekend

FelicitusNeko commented:
If nothing else, it builds in Linux. I can try to finagle building for Windows, then give this a run in one or the other.

Rnd-Guy commented:
Got my environment working I think, might still be scuffed though

However it looks like there's a bug when generating a dungeon that needs to regenerate due to my new crystal safety check.
It looks like it's not cleaning up the enemies sufficiently and it's causing enemies to appear outside walls and cause horde rooms to be impossible, as well as glitching up the palette

![image](https://github.com/user-attachments/assets/c644b124-510a-4f48-82af-0164b0c50b76)

To reproduce (I think), spam pressing new game until the loading bar noticeably resets (this means it's recreating the dungeon as it's not meeting some criteria which for small dungeons should only be my new min crystal check)

I've never seen this during all of my testing, so I dunno if this is due to my current scuffed environment or not but I'll do some more testing tomorrow

Rnd-Guy committed:
reset enemies when retrying generate

As enemies are now part of generate, we need to make sure we clean up failed attempts

Rnd-Guy commented:
Thankfully appears to be a simple fix, just needed to clean up enemies after each generate attempt

Successful 300 room run with latest version which reset gen a few times before starting, got 2 of the 17th caches which matches with my previous testing results when having 0 evo traps + no crystal efficiency
![image](https://github.com/user-attachments/assets/ea1542a1-a7c7-445f-923a-78a6404044c6)

Rnd-Guy commented:
Putting this as draft for a bit:
I've unfortunately borked my c/c++ build environment again and I can't tell if I'm getting bugs from that or if it's legitimate bugs in the code, last time I tried (which was a while ago) there seemed to be so many more issues than I remember T_T

Due to this being a large change and there being so much testing to be done to make sure it all works, it's hard to tell when this is ready to be merged.
I do think the vast majority of the work has been done, especially in terms of crystal changes and balancing, but it's really hard to test for stability without pouring hours into manual testing.

I wonder if it's a good idea to instead create beta builds for this piece of code (and a beta apworld) first before merging. Especially when it comes to testing multiple values of room counts

In any case, I'll leave this as draft until I find some more time to remind myself what the remaining issues for this are

FelicitusNeko commented:
I definitely had similar problems coming back to this recently. The biggest issue at the moment is that the current version of websocketpp is not compatible with the current version of asio, and the former has not been updated in a hot minute. As a result, it's necessary to either downgrade asio, hope websocketpp gets updated, or maybe look into adding asio as a submodule so we can use an old version and link it statically. But yeah, the problem is entirely upstream, which sucks.

Of course, I could have also merged this in a more timely manner in the first place, and then we wouldn't have this problem to begin with 😓

Rnd-Guy commented:
It's all good, this is a rather significant change and maintenance is always a lot of work! This is still something I would be happy to see finished though I won't have the time to look at it for a while 😭 

If you ever want to discuss this a bit more, do feel free to reach out to me! I think I should've discussed a lot of this with you before I started working on it ^^;

For now I think I'll try to annotate what I can remember about the code in this PR to explain the changes for you and future me

Rnd-Guy commented on src/apinterface.cpp:
```diff
@@ -269,6 +269,21 @@ void ConnectAP()
+      rooms_to_knife = (rooms_to_gen * agate_knife_percent) / 100;
+      if (rooms_to_knife < 1) rooms_to_knife = 1; // 0 will cause the knife to never spawn, 1 spawns it in the starting room
+    }
+    else rooms_to_knife = rooms_to_gen;
```
(decision)

I've been having second thoughts about whether allowing agate knife percent as an option is sensible haha
It's great on paper as it allows you to set this to like 95% so you don't have to find the small single missed rooms but it's really easy to set this to something lower and be insanely busted.
Having small room count already helps make the true ending much more accessible already after all.

This could also be moved to a separate PR if that is more convenient, it doesn't really get in the way of the room count stuff

Open to suggestions!

Rnd-Guy commented on src/apinterface.cpp:
```diff
@@ -283,6 +298,7 @@ void ConnectAP()
     ap_connect_sent = false;
     if (std::find(errors.begin(), errors.end(), "InvalidSlot") != errors.end()) {
       //bad_slot(game?game->get_slot():"");
+      SetAPStatus("InvalidSlot", 1);
```
(info)

Since we need to connect at the start (to load the room count set by the server), we need to identify when a connection has failed and abort loading into the game. This is then used to send the player back to the title screen.

Rnd-Guy commented on src/audio.c:
```diff
@@ -128,29 +128,29 @@ void BackgroundMusic()
 	Mix_VolumeMusic(96);
 	
-	if (rooms[player_room].s_dist <= 15) {
+	if (rooms[player_room].s_dist * dist_scaling <= 15) {
```
(info)

Change from hardcoded distance to instead be based on the actual max size of the dungeon

Rnd-Guy commented on src/demon.c:
```diff
@@ -644,6 +654,7 @@ struct enemy *CreateEnemyEx(int enemy_x, int enemy_y, int enemy_room, int enemy_
 	AddEnemyPos(new_enemy);
 
 	total_enemies++;
+	average_gems += (new_enemy->min_gems + new_enemy->max_gems) / 2.0f;
```
(info)

Debugging - This is used to gauge how much gems actually exist in the dungeon, by directly measuring the worth of every monster.

Not used in any calculations at the moment, but if we wanted to force the dungeon to have the required number of gems, then we can use this or perhaps the lower bound to ensure that there's always enough gems to purchase stuff

Rnd-Guy commented on src/demon.c:
```diff
@@ -688,12 +699,33 @@ void CreateGem(int x, int y, int r, int v)
+	fractional_crystals += float_value - (float)value;
+	int whole_crystals = (int)fractional_crystals;
+	value += whole_crystals;
+	fractional_crystals -= (float)whole_crystals;
```
(info)

Keep track of rounding errors by adding the remainder to the next crystal that dropped
Example:
Fractional crystals = 0, scaling is 1.4, two gems dropped with original value 2

1)
full value = 2 * 1.4 = 2.8
gem value  = 2
remaining fractional crystals = 0 + 0.8
final gem value = 2
final fractional crystals = 0.8

2)
full value = 2 * 1.4 = 2.8
gem value = 2
remaining fractional crystals = 0.8 + 0.8 = 1.6 = 1 + 0.6
final gem value = 2 + 1 = 3
final fractional crystals = 0.6

So we'll have a gem of value 2 and a gem of value 3 and we leave behind 0.6 for the next crystal that drops

Rnd-Guy commented on src/demon.c:
```diff
@@ -688,12 +699,33 @@ void CreateGem(int x, int y, int r, int v)
+	fractional_crystals -= (float)whole_crystals;
+
+	// just in case scaling accidentally makes a crystal hit the magic number for hearts
+	if (value == hp_gem_value) {
```
(info)

in case the multiplication actually hits the magic number 31337 as that will suddenly turn it into a heart instead. Don't think that can actually happen though

Rnd-Guy commented on src/demon.c:
```diff
@@ -941,7 +975,15 @@ void InitEnemies()
+	// however we cut off dungeons that gen with too fewer crystals, which bumps the average by a bit, so we'll use 10000 to tune it down
+	//    (eg. average of 8,9,10,11,12 and 6,8,10,12,14 are both 10 but if you remove values below 8, then average of the second set is now 11)
+	if (rooms_to_gen < 3000) crystal_scaling = 10000.0f / (float)total_enemies;
+	else crystal_scaling = 1.0f;
```
(discussion)

The main crystal scaling behaviour is done here. The current code is my second attempt at scaling gems but it's raised a lot of issues. Here's a list of some other ideas, I'm open to suggestions

Approach one: scale based on room count
* This failed badly as having 50% rooms still had about 70% enemies, so there were too many gems.

Approach two (as shown): scale based on monster count
* Keeps a rather natural/organic variance in terms of total gem counts
* Small dungeons are susceptible to extremely high variance. A countermeasure had to be used to prevent dungeons from being generated with too few gems but as a result generates more gems than a 3000 room dungeon.

Approach three: scale gems based on average gem count
* Can better control the variance, or sprinkle our own randomness in
* Can fully control the amount of gems as a result, eg to meet cache costs and prevent there being too few gems to buy logical caches.
* Less organic, may make dungeons feel samey

Approach four: scale gems based on minimum gem count
* Absolute control of the gem count, or sprinkle our own randomness in
* If we do this and remove permanently losing gems on death then we can guarantee that there is always enough gems for cache purchases.

Rnd-Guy commented on src/demon.c:
```diff
@@ -1000,8 +1042,9 @@ void ActivateEnemies(int room)
-		max_activate_dist = rooms[room].s_dist;
+	// max_activate_dist is used to determine the enemy types that can be activated, the higher it is, the stronger the enemy can be
+	if ((int)(rooms[room].s_dist * dist_scaling) > max_activate_dist) {
+		max_activate_dist = rooms[room].s_dist * dist_scaling;
```
TODO: max_activate_dist probably needs to be scaled as well prior to the if condition

Rnd-Guy commented on src/demon.c:
```diff
@@ -1216,7 +1259,7 @@ void ArtifactRoomUnlock(int room)
 	// place treasure
 	placed = 0;
-	tot_treasures = 2 + rand() % (rooms[room].s_dist / 8 + 1);
+	tot_treasures = 2 + rand() % (int)((rooms[room].s_dist * dist_scaling) / 8 + 1);
```
(info)

Further treasure rooms drop more chests, this should keep them to be proportionally similar to 3000 room behaviour

Rnd-Guy commented on src/demon.c:
```diff
@@ -1534,7 +1577,7 @@ void MoveEnemy(struct enemy *e)
 						// add_int_stat(STAT_BULLETS_CANCELLED, 1);
 						b->delete_me = 1;
-						CreateGem(b->x, b->y, b->room,
+						CreateEnemyGem(b->x, b->y, b->room,
```
(info)

Gems are dropped either on enemy death or player death.
We only want to scale gem values on enemy death.

Rnd-Guy commented on src/ending.c:
```diff
@@ -365,11 +366,16 @@ char *credits[] = {
   draw_text_f(66, 30, "Circuit bursts     %9d", 192, get_int_stat(STAT_BURSTS));
   draw_text_f(66, 40, "Successful hits    %9d", 168, get_int_stat(STAT_HITS));
-  draw_text_f(66, 50, "Kills              %9d", 192, get_int_stat(STAT_KILLS));
+  draw_text_f(66, 50, "Kills            %11s", 192, kills);
```
(info)

show kills over total enemies eg `10101/10200`

Rnd-Guy commented on src/ending.c:
```diff
@@ -394,7 +400,12 @@ void DrawStats()
+  else {
+    char explored_fraction[10] = { '\0' };
+    sprintf(explored_fraction, "%d/%d", explored, rooms_to_gen);
+    draw_text_f(350, 130, "Rooms explored     %9s", 168, explored_fraction);
```
(info)

show rooms over total rooms eg `123/300`

Rnd-Guy commented on src/ending.c:
```diff
@@ -410,6 +421,12 @@ void DrawStats()
+    draw_text_f(20, 280, "Debug stats:", 192);
+    draw_text_f(20, 290, "Average Gems Left  %9.1f", 168, average_gems);
+    draw_text_f(20, 300, "^ After Multiplier %9.1f", 192, expected_total_gems);
+    draw_text_f(350, 290, "Max room distance  %9d", 168, max_dist);
```
(jnfo)

debug stats - estimate gem count for dungeon without needing to clear it manually

Rnd-Guy commented on src/levelblit.c:
```diff
@@ -799,6 +799,26 @@ void SavingScreen(int part, float progress)
+
+  VideoUpdate();
+  DummyEventPoll();
+}
```
(info)

Connection screen that now appears before the game is started or continued. We need to get the room count from the server so we add a box to show connection status which will tell us if the connection has failed

Rnd-Guy commented on src/mapgen.c:
```diff
@@ -976,12 +1009,33 @@ int Generate()
+  if (expected_total_gems < 250000.0f) {
+    DoRepeat = 0;
+    ResetLevel();
+    return 0;
```
(info)

Currently small dungeons have too much variance and can generate with significantly fewer crystals. See [this comment](https://github.com/FelicitusNeko/meritous-ap/pull/24/files#diff-903157efb537a1eeeb784ea7feca375c968455685310992c270713517af8973aR983) for why.

This is a workaround that forces the generation to abort and restart if not enough gems exist. It doesn't take long for the dungeon to regenerate

Rnd-Guy commented on src/levelblit.c:
```diff
@@ -880,13 +900,32 @@ int DungeonPlay(const char *fname)
+  }
+
+  LoadingScreen(1, 0.0);
+
```
(info)

Connection is now required prior to generating the dungeon as we now want to read the room count from the server.
I don't remember why InitStores needed to be moved as well anymore but I remember it was having a lot of issues. 

TODO why init stores here?

Rnd-Guy commented on src/levelblit.c:
```diff
@@ -2885,3 +2922,38 @@ void ThinLine(SDL_Surface *scr, int x1, int y1, int x2, int y2, Uint8 col)
 }
 
+// waits until a successful connection is made to the archipelago server. Requires stores to be initialised beforehand
+int WaitForConnection() {
```
(info)

This is the new connection dialog that happens just before the game is loaded, where we wait until we get the room count from the server.

Rnd-Guy commented on src/mapgen.c:
```diff
@@ -41,6 +42,16 @@ int DoRepeat = 0;
+// vanilla does calculations based on a max dist of 50, but can gen dungeons with higher than 50.
+// we then mimic this overshooting for small dungeons too, to try to keep the proportion of difficult/far rooms the same.
+// generated 50 vanilla worlds and got their average max_dist (51.4) so for calculations we'll use 50/51.4 (ie 97.276%) of the actual max distance
+const float max_dist_overshoot = 0.97276f;
```
(info)

This is an attempt to keep the proportions of enemies in rooms similar to current behaviour, for both gameplay reasons and for gem scaling reasons.

Rnd-Guy commented on src/levelblit.c:
```diff
@@ -880,13 +900,32 @@ int DungeonPlay(const char *fname)
   }
 
   RandomGenerateMap();
-  InitEnemies();
```
(info)

Enemies are now part of generation and may be the reason for a dungeon to regenerate (in case too few gems have spawned). As a result it has now moved to be inside `RandomGenerateMap`

Note if we do decide to change the way crystals are scaled, then this can be reverted

Rnd-Guy commented on src/mapgen.c:
```diff
@@ -221,13 +246,14 @@ void RandomGenerateMap()
   if (game_load) {
     NewLevel();
     ReadMapData();
+    InitEnemies();
```
(info)

Generating enemies are now part of `Generate`, so here would be the equivalent for loading the game

Rnd-Guy commented on src/mapgen.c:
```diff
@@ -221,13 +246,14 @@ void RandomGenerateMap()
+    InitEnemies();
   } else {
     NewLevel();
+    boss_room_interval = rooms_to_gen / 6;
```
(info)

If rooms to gen is extremely low (about <50), it's possible that rounding errors cause there to be 7 boss rooms. No issues arise from this but I was surprised when I saw it
